### PR TITLE
New Console Logger Test + Discard before Eval

### PR DESCRIPTION
**Pull request by rishab-partha**
# What does this PR do?

New PR for the console logger test, which is now smaller and has documentation, for the discard before eval and not at the end of epoch functionality.

# What issue(s) does this change relate to?

[CO-2152](https://mosaicml.atlassian.net/browse/CO-2152)
[CO-2030](https://mosaicml.atlassian.net/browse/CO-2030
)

[CO-2152]: https://mosaicml.atlassian.net/browse/CO-2152?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[CO-2030]: https://mosaicml.atlassian.net/browse/CO-2030?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by mvpatel2000** [comment]
@eracah can you take a look at test?

**Review by mvpatel2000 — COMMENTED** [review]
lgtm, will wait for evan's sign off

**Comment by rishab-partha** [comment]
Added support for flushing the logger at FIT_END iff the logger was not flushed the same batch (otherwise, the user sees two identical logs, which is very confusing).

Also, updated tests to match.

File: composer/loggers/console_logger.py [hunk 1]
```diff
@@ -96,21 +98,26 @@ def epoch_end(self, state: State, logger: Logger) -> None:
...
+            self.last_log = cur_batch
             # Clear logged metrics.
             self.logged_metrics = {}
 
     def fit_end(self, state: State, logger: Logger) -> None:
         # Always clear logged metrics so they don't get logged in a subsequent eval call.
+        cur_batch = int(state.timestamp.batch)
+        if self.last_log != cur_batch:
+            self.log_to_console(self.logged_metrics, prefix='Train ',
+                                state=state)  # log at the end of training if you didn't just log
```
Comment (mvpatel2000) [inline composer/loggers/console_logger.py:119]: can we get rid of `self.last_log` and instead check if `self.logged_metrics` is non-empty, log it?

File: tests/loggers/test_console_logger.py [hunk 2]
```diff
@@ -56,6 +57,8 @@ def test_console_logger_interval(console_logger_test_stream, console_logger_test
     with open(console_logger_test_file_path, 'r') as f:
         lines = f.readlines()
 
+    for line in lines:
+        print(line)
```
Comment (mvpatel2000) [inline tests/loggers/test_console_logger.py:61]:
```suggestion
```

**Comment by rishab-partha** [comment]
Changed the behavior to only print at fit-end only if the last batch did not log. There are longer-term QoL improvements that should be made in the future.

File: composer/loggers/console_logger.py [hunk 3]
```diff
@@ -96,21 +98,26 @@ def epoch_end(self, state: State, logger: Logger) -> None:
 
         if unit == TimeUnit.EPOCH and (cur_epoch % int(self.log_interval) == 0 or cur_epoch == 1):
             self.log_to_console(self.logged_metrics, prefix='Train ', state=state)
-        # Always clear logged metrics so they don't get logged in a subsequent eval call. The
-        # metrics will be recomputed and overridden in future batches so they can be safely
-        # discarded.
-        self.logged_metrics = {}
+            self.last_logged_batch = int(state.timestamp.batch)
+            # Clear logged metrics
+            self.logged_metrics = {}
```
Comment (mvpatel2000) [inline composer/loggers/console_logger.py:103]:
```suggestion
            self.logged_metrics = {}. # Clear logged metrics
```
nit

**Review by mvpatel2000 — APPROVED** [review]

File: tests/loggers/test_console_logger.py [hunk 4]
```diff
@@ -287,3 +302,57 @@ def test_console_logger_with_a_callback(console_logger_test_stream, console_logg
...
+    actual_num_log_lines = sum(
+        [1 if bool(reg_exp.search(line)) and ('trainer/' not in line and 'time/' not in line) else 0 for line in lines])
+
+    assert model.train_metrics is not None
+    num_metrics = len(list(model.train_metrics.keys())) if isinstance(model.train_metrics, MetricCollection) else 1
+    num_metrics += len(list(model.parameters())) * 7 + 1  # number from Adam
+
+    num_losses = 1
+    num_metrics_and_losses_per_logging_event = num_metrics + num_losses
+
```
Comment (eracah) [inline tests/loggers/test_console_logger.py:354]: add a comment explaining why it's this formula

File: tests/loggers/test_console_logger.py [hunk 5]
```diff
@@ -287,3 +302,57 @@ def test_console_logger_with_a_callback(console_logger_test_stream, console_logg
...
+    # Make a regular expression for matches for any line that contains "Train" followed by
+    # a colon.
+    reg_exp = re.compile('Train *:*')
+    actual_num_log_lines = sum(
+        [1 if bool(reg_exp.search(line)) and ('trainer/' not in line and 'time/' not in line) else 0 for line in lines])
+
+    assert model.train_metrics is not None
+    num_metrics = len(list(model.train_metrics.keys())) if isinstance(model.train_metrics, MetricCollection) else 1
+    num_metrics += len(list(model.parameters())) * 7 + 1  # number from Adam
```
Comment (eracah) [inline tests/loggers/test_console_logger.py:350]: explain that it's 7 metrics per layer

**Review by eracah — APPROVED** [review]
LGTM

**Pull request merged**